src/typing/constraint_js.ml

```diff
@@ -529,21 +529,34 @@ let copy_node node = match node with
     Root { rank; constraints = Unresolved (copy_bounds bounds) }
   | _ -> node
 
+type scope_kind =
+  | VarScope (* var, functions hoisted up to this point *)
+  | LexicalScope (* let, const, classes *)
+
 type scope_entry = {
   specific: Type.t;
   general: Type.t;
   def_loc: Spider_monkey_ast.Loc.t option;
   for_type: bool;
+  constant: bool;
+  scope_kind: scope_kind;
```

Comment:
Reading top-down, so apologies if the motivation becomes clear later, but: wouldn't the scope_kind for an entry be completely determined by the scope the entry was in?